### PR TITLE
fix(agw): Migrated ue_id_coll hashlist to protobuf map

### PR DESCRIPTION
**Pull request by rsarwad**
## Summary
Migrated ue_id_coll hashlist to protobuf map. The PR partially addresses the issue, https://github.com/magma/magma/issues/11190

All tasks are initialized from oai_mme.c file in main function. As we port the individual tasks to cpp, initialize functions are called from main function. And declaration of initialization function of all tasks shall we moved to include/mme_init.hpp file
## Test Plan
Executed s1ap sanity test suite

### CODE REVIEW

File: lte/gateway/c/core/oai/include/mme_init.hpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
+/**
+ * Copyright 2022 The Magma Authors.
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/include/mme_init.hpp:2]: What's the purpose of creating this new file? 

Further, this is initializing the S1AP task, not the MME task. So the name is misleading.

Comment (rsarwad) [reply lte/gateway/c/core/oai/include/mme_init.hpp:2]: 1. All tasks are initialized from oai_mme.c file in main function. 
2. As we port the individual tasks to cpp, initialize functions are called from main function. And all these initialize shall be put in this file.
3. Since we started with s1ap task, right now only s1ap task's functionalities are available

File: lte/gateway/c/core/oai/tasks/s1ap/s1ap_state_converter.cpp [hunk 2]
```diff
@@ -145,21 +146,17 @@ void S1apStateConverter::proto_to_enb(const oai::EnbDescription& proto,
...
 
   // load ues
-  hashtable_rc_t ht_rc;
-  auto ht_name = bfromcstr("s1ap_ue_coll");
-
-  hashtable_uint64_ts_init(&enb->ue_id_coll, mme_config.max_ues, nullptr,
-                           ht_name);
-  bdestroy(ht_name);
+  proto_map_rc_t rc = {PROTO_MAP_OK};
+  enb->ue_id_coll.map = new google::protobuf::Map<uint32_t, uint64_t>();
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/s1ap/s1ap_state_converter.cpp:150]: The proto map is using a uint64 key. Can we use the same width here?

File: lte/gateway/c/core/oai/tasks/s1ap/s1ap_state_converter.cpp [hunk 3]
```diff
@@ -122,7 +122,8 @@ void S1apStateConverter::enb_to_proto(enb_description_t* enb,
   proto->set_ran_cp_ipaddr_sz(enb->ran_cp_ipaddr_sz);
 
   // store ue_ids
-  hashtable_uint64_ts_to_proto(&enb->ue_id_coll, proto->mutable_ue_ids());
+  proto_map_uint32_uint64_to_proto(enb->ue_id_coll, proto->mutable_ue_ids());
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/s1ap/s1ap_state_converter.cpp]: Can we just pass the enb->ue_id_coll reference to initialize the proto field? Eventually, this conversion should not be needed as the enb->ue_id_coll should be directly serialized to S1ap state.

**Review by ssanadhya — CHANGES_REQUESTED** [review]

File: lte/gateway/c/core/oai/include/mme_init.hpp (hunk 1 above)
Comment (ssanadhya) [reply lte/gateway/c/core/oai/include/mme_init.hpp:2]: Can you add this to the PR description and also the parent issue to guide the migration strategy ? Also, is the goal to eventually get rid of this file?

Comment (rsarwad) [reply lte/gateway/c/core/oai/include/mme_init.hpp:2]: Updated the comment. Planning to retain this file as declaration of initialization function of all tasks shall be part of this file.

File: lte/gateway/c/core/oai/tasks/s1ap/s1ap_state_converter.cpp (hunk 2 above)
Comment (rsarwad) [reply lte/gateway/c/core/oai/tasks/s1ap/s1ap_state_converter.cpp:150]: We discussed that, since mme_ue_s1ap_if type uint32 and shall change the datatype in proto file.

File: lte/gateway/c/core/oai/tasks/grpc_service/S1apServiceImpl.cpp [hunk 4]
```diff
@@ -14,14 +14,14 @@
  * For more information about the OpenAirInterface (OAI) Software Alliance:
  *      contact@openairinterface.org
  */
-#include "lte/gateway/c/core/oai/tasks/grpc_service/S1apServiceImpl.hpp"
 
 #include <string>
+#include "lte/gateway/c/core/oai/tasks/grpc_service/S1apServiceImpl.hpp"
```
Comment (ajahl) [inline lte/gateway/c/core/oai/tasks/grpc_service/S1apServiceImpl.cpp:18]: Why did you change the order? The include of the own interface should be on the first position.

**Review by nstng — APPROVED** [review]
bazel changes lgtm (only BUILD.bazel change)

File: lte/protos/oai/s1ap_state.proto [hunk 5]
```diff
@@ -34,7 +34,7 @@ message EnbDescription {
   uint32 instreams = 8;        // sctp_stream_id_t
   uint32 outstreams = 9;      // sctp_stream_id_t
 
-  map<uint64, uint64> ue_ids = 10; // mme_ue_s1ap_id -> comp_s1ap_id
```
Comment (ssanadhya) [inline lte/protos/oai/s1ap_state.proto:37]: You cannot modify the proto field since it will not be backwards compatible. We need to mark this as deprecated, and add a new field 
``` 
map<uint32, uint64> ue_id_map = 14;
```

**Review by ssanadhya — CHANGES_REQUESTED** [review]
This version is still using `proto_map_uint32_uint64_to_proto` to copy the in-memory object to the proto generated object. Do we still need it after changing `s1ap_state.proto`?

File: lte/protos/oai/s1ap_state.proto (hunk 5 above)
Comment (rsarwad) [reply lte/protos/oai/s1ap_state.proto:37]: Sure.
Is this the way we mark unused field as deprecated, https://github.com/magma/magma/blob/master/lte/protos/oai/s1ap_state.proto#L8?

**Comment by rsarwad** [comment]
> This version is still using `proto_map_uint32_uint64_to_proto` to copy the in-memory object to the proto generated object. Do we still need it after changing `s1ap_state.proto`?
You are right. We will not be using proto_map_uint32_uint64_to_proto to copy the in-memory object to the proto generated object.  My mistake, I had changes in another branch but forgot take those changes here.
Thanks for letting me know.

File: lte/gateway/c/core/oai/include/s1ap_types.hpp [hunk 6]
```diff
@@ -164,8 +165,9 @@ typedef struct enb_description_s {
   /** UE list for this eNB **/
   /*@{*/
   uint32_t nb_ue_associated;  ///< Number of NAS associated UE on this eNB
-  hash_table_uint64_ts_t ue_id_coll;  ///< Contains comp_s1ap_id assoc to
-                                      ///< enodeb, key is mme_ue_s1ap_id;
+  magma::proto_map_uint32_uint64_t
+      ue_id_coll;  ///< Contains comp_s1ap_id assoc to
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/include/s1ap_types.hpp]: The comment here can be simplified to

```
/// key: mme_ue_s1ap_id, value: comp_s1ap_id
```

File: lte/gateway/c/core/oai/include/mme_init.hpp (hunk 1 above)
Comment (ssanadhya) [reply lte/gateway/c/core/oai/include/mme_init.hpp:2]: But the init function for each task should be part of that task directory, instead of being in a common header. It will be good to retain this under s1ap_mme.hpp.

File: lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_handlers.cpp [hunk 7]
```diff
@@ -560,22 +558,22 @@ status_code_e s1ap_mme_handle_s1_setup_request(s1ap_state_t* state,
...
+        enb_association->nb_ue_associated, enb_association->ue_id_coll.size());
     rc = s1ap_mme_generate_s1_setup_failure(
         assoc_id, S1ap_Cause_PR_transport,
         S1ap_CauseTransport_transport_resource_unavailable,
         S1ap_TimeToWait_v20s);
     increment_counter("s1_setup", 1, 2, "result", "failure", "cause",
                       "invalid_state");
     /* UE state at s1ap task is created on reception of initial ue message
-     * Hash list, ue_id_coll is updated after mme_app_task assigns and provides
+     * map, ue_id_coll is updated after mme_app_task assigns and provides
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_handlers.cpp]:
```suggestion
     * The map ue_id_coll is updated after mme_app_task assigns and provides
```

File: lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_handlers.cpp [hunk 8]
```diff
@@ -3704,12 +3699,12 @@ status_code_e s1ap_handle_sctp_disconnection(s1ap_state_t* state,
 
   if (reset) {
     /* UE state at s1ap task is created on reception of initial ue message
-     * Hash list, ue_id_coll is updated after mme_app_task assigns and provides
+     * map, ue_id_coll is updated after mme_app_task assigns and provides
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_handlers.cpp]:
```suggestion
     * The map ue_id_coll is updated after mme_app_task assigns and provides
```

**Comment by rsarwad** [comment]
> proto_map_uint32_uint64_to_proto



> This version is still using `proto_map_uint32_uint64_to_proto` to copy the in-memory object to the proto generated object. Do we still need it after changing `s1ap_state.proto`?

Incorporated review comment.

**Review by ssanadhya — COMMENTED** [review]
Minor comments, will continue reviewing

File: lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_handlers.cpp [hunk 9]
```diff
@@ -3704,12 +3699,12 @@ status_code_e s1ap_handle_sctp_disconnection(s1ap_state_t* state,
 
   if (reset) {
     /* UE state at s1ap task is created on reception of initial ue message
-     * Hash list, ue_id_coll is updated after mme_app_task assigns and provides
+     * The map, ue_id_coll is updated after mme_app_task assigns and provides
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_handlers.cpp]:
```suggestion
     * The map ue_id_coll is updated after mme_app_task assigns and provides
```

**Review by ssanadhya — APPROVED** [review]
Lgtm, with minor nit
